### PR TITLE
Document finalizer task-switch restriction

### PR DESCRIPTION
**Pull request by ihnorton**
ref https://github.com/JuliaLang/julia/issues/21442
and https://discourse.julialang.org/t/let-expression-and-garbage-collection/7804

### CODE REVIEW

**Comment by thofma** [comment]
Could the hint about `@schedule` be expanded? Maybe an easy example illustrating how to print something. I think this is a very common case.

**Comment by ihnorton** [comment]
Sure, done.

**Comment by nstiurca** [comment]
It looks like `@schedule` is deprecated now, so the hint should probably be to use `@async`

File: base/gcutils.jl [hunk 1]
```diff
@@ -10,6 +10,10 @@
 Register a function `f(x)` to be called when there are no program-accessible references to
 `x`, and return `x`. The type of `x` must be a `mutable struct`, otherwise the behavior of
 this function is unpredictable.
+
+`f` must not cause a task switch, which excludes most I/O operations such as `println`.
+`@schedule println("message")` or `ccall(:jl_, Void, (Any,), "message")` may be helpful for
```
Comment (nstiurca) [inline base/gcutils.jl]: It looks like `@schedule` is deprecated now, so the hint should probably be to use `@async`

**Comment by iamed2** [comment]
Wow, thank you for documenting this, I had no idea

**Comment by tkf** [comment]
I suppose `Core.println` and `Core.print` can be used in finalizer too?

**Comment by antoine-levitt** [comment]
This is a pretty easy limitation to run into when debugging, and it's not obvious that IO task switching is the issue. Can this PR be merged?

**Comment by iamed2** [comment]
Had to ask about this on Slack, this should really be in the documentation by now

**Comment by ihnorton** [comment]
I can no longer update the branch, so please feel free to edit/merge as you see fit.

**Pull request merged**